File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int[] dfsHelper(TreeNode root, int[] goodNodes, int maxTillNow) {
        if(root == null) return goodNodes;
        if(root.val >= maxTillNow) {
            maxTillNow = root.val;
            goodNodes[0]++;
        }
        dfsHelper(root.left, goodNodes, maxTillNow);
        dfsHelper(root.right, goodNodes, maxTillNow);
        return goodNodes;
    }
    public int goodNodes(TreeNode root) {
        int[] goodGuys = {0};
        int[] ans = dfsHelper(root, goodGuys, Integer.MIN_VALUE);
        return ans[0];
    }
}

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int helper(TreeNode* root, int &gdNodes, int maxTillNow) {
        if(!root) return gdNodes;
        if(root->val >= maxTillNow) {
            maxTillNow = root->val;
            gdNodes++;
        }
        helper(root->left, gdNodes, maxTillNow);
        helper(root->right, gdNodes, maxTillNow);
        return gdNodes;
    }
public:
    int goodNodes(TreeNode* root) {
        if(!root) return 0;
        int gdNodes = 0, maxTillNow = INT_MIN;
        return helper(root, gdNodes, maxTillNow);
    }
};